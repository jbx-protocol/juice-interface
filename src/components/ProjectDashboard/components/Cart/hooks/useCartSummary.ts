import { useProjectCart } from 'components/ProjectDashboard/hooks'
import { NftRewardsContext } from 'contexts/NftRewards/NftRewardsContext'
import { useWallet } from 'hooks/Wallet'
import { useCallback, useContext, useMemo } from 'react'
import { formatCurrencyAmount } from 'utils/formatCurrencyAmount'

export const useCartSummary = () => {
  const { isConnected, connect } = useWallet()
  const cart = useProjectCart()
  const nftRewardsCtx = useContext(NftRewardsContext).nftRewards
  const tiers = useMemo(
    () => nftRewardsCtx.rewardTiers ?? [],
    [nftRewardsCtx.rewardTiers],
  )
  const nftRewards = useMemo(() => {
    return cart.nftRewards
      .map(nft => ({ ...tiers.find(tier => tier.id === nft.id), ...nft }))
      .filter(nft => nft !== undefined)
      .map(nft => ({
        fileUrl: nft.fileUrl,
        name: nft.name,
        quantity: nft.quantity,
        id: nft.id,
      }))
  }, [cart.nftRewards, tiers])

  const amountText = cart.totalAmount
    ? formatCurrencyAmount(cart.totalAmount)
    : undefined

  const resetCart = useCallback(() => {
    cart.dispatch({ type: 'reset' })
  }, [cart])

  const payProject = useCallback(() => {
    if (!isConnected) {
      connect()
      return
    }
    cart.dispatch({ type: 'openPayModal' })
  }, [cart, connect, isConnected])

  const showCurrencyOnCollapse = useMemo(() => {
    return cart.payAmount?.amount ? cart.payAmount.amount > 0 : false
  }, [cart.payAmount])

  return {
    amountText,
    currency: cart.totalAmount?.currency,
    nftRewards,
    walletConnected: isConnected,
    showCurrencyOnCollapse,
    resetCart,
    payProject,
  }
}
