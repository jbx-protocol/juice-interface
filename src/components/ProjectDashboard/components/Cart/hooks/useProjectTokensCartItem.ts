import {
  useProjectCart,
  useTokensPerEth,
} from 'components/ProjectDashboard/hooks'
import { useCallback } from 'react'

export const useProjectTokensCartItem = () => {
  const {
    totalAmount: payAmount,
    userIsReceivingTokens,
    dispatch,
  } = useProjectCart()
  const { receivedTickets } = useTokensPerEth(payAmount)

  const removeTokens = useCallback(
    () => dispatch({ type: 'removeTokens' }),
    [dispatch],
  )
  return {
    userIsReceivingTokens,
    receivedTickets,
    removeTokens,
  }
}
