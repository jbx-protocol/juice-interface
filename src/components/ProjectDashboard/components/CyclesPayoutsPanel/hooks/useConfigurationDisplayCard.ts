import { t } from '@lingui/macro'
import { useMemo } from 'react'

export const useConfigurationDisplayCard = (type: 'current' | 'upcoming') => {
  const title = useMemo(() => {
    if (type === 'current') {
      return t`Current cycle`
    }
    return t`Upcoming cycle`
  }, [type])

  return {
    title,
  }
}
