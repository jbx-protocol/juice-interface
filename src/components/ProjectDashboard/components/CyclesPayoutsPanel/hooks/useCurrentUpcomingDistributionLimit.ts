import {
  useProjectContext,
  useProjectMetadata,
} from 'components/ProjectDashboard/hooks'
import useProjectDistributionLimit from 'hooks/v2v3/contractReader/useProjectDistributionLimit'
import { useProjectUpcomingFundingCycle } from 'hooks/v2v3/contractReader/useProjectUpcomingFundingCycle'

export const useCurrentUpcomingDistributionLimit = (
  type: 'current' | 'upcoming',
) => {
  const { projectId } = useProjectMetadata()
  const {
    primaryETHTerminal,
    distributionLimit: currentDistributionLimit,
    distributionLimitCurrency: currentDistributionLimitCurrency,
  } = useProjectContext()

  const {
    data: upcomingFundingCycleData,
    loading: upcomingFundingCycleLoading,
  } = useProjectUpcomingFundingCycle({ projectId })
  const [upcomingFundingCycle] = upcomingFundingCycleData ?? []

  const {
    data: upcomingDistributionLimitData,
    loading: upcomingDistributionLimitLoading,
  } = useProjectDistributionLimit({
    projectId,
    configuration: upcomingFundingCycle?.configuration.toString(),
    terminal: primaryETHTerminal,
  })

  const [upcomingDistributionLimit, upcomingDistributionLimitCurrency] =
    upcomingDistributionLimitData ?? []

  if (type === 'current') {
    return {
      distributionLimit: currentDistributionLimit,
      distributionLimitCurrency: currentDistributionLimitCurrency,
      loading: false,
    }
  }
  return {
    distributionLimit: upcomingDistributionLimit,
    distributionLimitCurrency: upcomingDistributionLimitCurrency,
    loading: upcomingFundingCycleLoading || upcomingDistributionLimitLoading,
  }
}
