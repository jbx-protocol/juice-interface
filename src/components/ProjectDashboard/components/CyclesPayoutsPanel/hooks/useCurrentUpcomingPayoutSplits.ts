import {
  useProjectContext,
  useProjectMetadata,
} from 'components/ProjectDashboard/hooks'
import { ETH_PAYOUT_SPLIT_GROUP } from 'constants/splits'
import useProjectSplits from 'hooks/v2v3/contractReader/useProjectSplits'
import { useProjectUpcomingFundingCycle } from 'hooks/v2v3/contractReader/useProjectUpcomingFundingCycle'

export const useCurrentUpcomingPayoutSplits = (
  type: 'current' | 'upcoming',
) => {
  const { projectId } = useProjectMetadata()
  const { payoutSplits: currentPayoutSplits } = useProjectContext()

  const {
    data: upcomingFundingCycleData,
    loading: upcomingFundingCycleLoading,
  } = useProjectUpcomingFundingCycle({ projectId })
  const [upcomingFundingCycle] = upcomingFundingCycleData ?? []

  const { data: upcomingPayoutSplits, loading: upcomingProjectSplitsLoading } =
    useProjectSplits({
      projectId,
      splitGroup: ETH_PAYOUT_SPLIT_GROUP,
      domain: upcomingFundingCycle?.configuration?.toString(),
    })

  if (type === 'current') {
    return { splits: currentPayoutSplits, loading: false }
  }
  return {
    splits: upcomingPayoutSplits,
    loading: upcomingFundingCycleLoading || upcomingProjectSplitsLoading,
  }
}
