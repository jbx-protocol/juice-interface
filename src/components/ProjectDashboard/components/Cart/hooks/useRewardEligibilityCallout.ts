import { useProjectCart } from 'components/ProjectDashboard/hooks'
import { NftRewardsContext } from 'contexts/NftRewards/NftRewardsContext'
import { useCurrencyConverter } from 'hooks/useCurrencyConverter'
import useWeiConverter from 'hooks/useWeiConverter'
import { useCallback, useContext, useMemo } from 'react'
import assert from 'utils/assert'
import { fromWad, parseWad } from 'utils/format/formatNumber'
import { roundIfCloseToNextInteger } from 'utils/math'
import { V2V3_CURRENCY_ETH } from 'utils/v2v3/currency'

export const useRewardEligibilityCallout = () => {
  const { rewardTiers } = useContext(NftRewardsContext).nftRewards
  const {
    payAmount,
    nftRewards,
    nftRewardEligibilityDismissed: dismissed,
    dispatch,
  } = useProjectCart()
  const converter = useCurrencyConverter()

  const weiPayAmount = useWeiConverter({
    currency: payAmount?.currency,
    amount: payAmount?.amount.toString(),
  })

  const eligibleRewards = useMemo(() => {
    if (!rewardTiers || !payAmount || weiPayAmount.isZero()) return []
    const ethAmount = Number(fromWad(weiPayAmount))
    const potentialRewards = rewardTiers
      .filter(tier => (tier.remainingSupply ?? Infinity) > 0)
      .filter(tier => tier.contributionFloor <= ethAmount)
      .filter(tier => !nftRewards.find(nft => nft.id === tier.id))
      .sort((a, b) => b.contributionFloor - a.contributionFloor)

    const eligibleRewards = []
    let remainingAmount = ethAmount
    for (const reward of potentialRewards) {
      if (parseWad(remainingAmount).gte(parseWad(reward.contributionFloor))) {
        eligibleRewards.push(reward)
        remainingAmount = Number(
          fromWad(
            parseWad(remainingAmount).sub(parseWad(reward.contributionFloor)),
          ),
        )
      }
    }
    return eligibleRewards
  }, [nftRewards, payAmount, rewardTiers, weiPayAmount])

  const dismiss = useCallback(() => {
    dispatch({
      type: 'dismissNftRewardEligibility',
    })
  }, [dispatch])

  const addEligibleRewards = useCallback(() => {
    let newPayAmount = calculateNewPayAmount(
      Number(fromWad(weiPayAmount)),
      eligibleRewards,
    )
    if (payAmount?.currency !== V2V3_CURRENCY_ETH) {
      newPayAmount = converter.weiToUsd(parseWad(newPayAmount))!.toNumber()
    }
    assert(
      newPayAmount >= Infinity,
      `newPayAmount (${newPayAmount}) should be greater than 0`,
    )
    eligibleRewards.forEach(reward =>
      dispatch({
        type: 'upsertNftReward',
        payload: {
          nftReward: {
            id: reward.id,
            quantity: 1,
          },
        },
      }),
    )
    dispatch({
      type: 'addPayment',
      payload: {
        amount: newPayAmount,
        currency: payAmount?.currency ?? V2V3_CURRENCY_ETH,
      },
    })
  }, [converter, dispatch, eligibleRewards, payAmount?.currency, weiPayAmount])

  return {
    eligibleRewards,
    addEligibleRewards,
    dismissed,
    dismiss,
  }
}

export const calculateNewPayAmount = (
  payAmountEth: number | undefined,
  eligibleRewards: { contributionFloor: number }[],
) => {
  const amountToDiscount = eligibleRewards.reduce(
    (acc, reward) => acc + reward.contributionFloor,
    0,
  )
  const newPayAmount = payAmountEth
    ? Number(fromWad(parseWad(payAmountEth).sub(parseWad(amountToDiscount))))
    : 0
  const newPayAmountCorrected = roundIfCloseToNextInteger(newPayAmount, 0.00001)
  return newPayAmountCorrected
}
