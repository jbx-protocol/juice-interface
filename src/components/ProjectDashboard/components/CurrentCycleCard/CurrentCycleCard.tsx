import { QuestionMarkCircleIcon } from '@heroicons/react/24/outline'
import { LockClosedIcon, LockOpenIcon } from '@heroicons/react/24/solid'
import { Trans, t } from '@lingui/macro'
import { Tooltip } from 'antd'
import { useCurrentCycleCard } from 'components/ProjectDashboard/hooks'
import { useProjectPageQueries } from 'components/ProjectDashboard/hooks/useProjectPageQueries'
import { TruncatedText } from 'components/TruncatedText'
import { useCallback, useMemo } from 'react'
import { twMerge } from 'tailwind-merge'
import { cycleTooltip } from '../CyclesPayoutsPanel/components/CyclesPanelTooltips'
import { DisplayCard } from '../ui'

export const CurrentCycleCard = ({ className }: { className?: string }) => {
  const { currentCycleNumber, timeRemainingText, isUnlocked } =
    useCurrentCycleCard()
  const { setProjectPageTab } = useProjectPageQueries()

  const openCyclePayoutsTab = useCallback(
    () => setProjectPageTab('cycle_payouts'),
    [setProjectPageTab],
  )

  const LockIcon = useMemo(
    () => (isUnlocked ? LockOpenIcon : LockClosedIcon),
    [isUnlocked],
  )

  const text = useMemo(
    () => (isUnlocked ? t`Unlocked` : timeRemainingText),
    [isUnlocked, timeRemainingText],
  )

  return (
    <DisplayCard
      className={twMerge('min-w-0 cursor-pointer pr-9', className)}
      onClick={openCyclePayoutsTab}
    >
      <Tooltip className="inline-flex items-center gap-1" title={cycleTooltip}>
        <span className="truncate whitespace-nowrap text-base font-medium">
          <Trans>Current Cycle: #{currentCycleNumber}</Trans>
        </span>
        <QuestionMarkCircleIcon className="h-4 w-4 text-grey-500 dark:text-slate-200" />
      </Tooltip>
      <div className="mt-4 flex min-w-0 items-center gap-2">
        <TruncatedText className="text-2xl font-medium" text={text} />
        <LockIcon className="h-4 w-4 flex-shrink-0 text-smoke-400 dark:text-slate-300" />
      </div>
      <TruncatedText
        className="mt-2 text-smoke-500 dark:text-slate-200"
        text={t`until Cycle #${currentCycleNumber + 1}`}
      />
    </DisplayCard>
  )
}
