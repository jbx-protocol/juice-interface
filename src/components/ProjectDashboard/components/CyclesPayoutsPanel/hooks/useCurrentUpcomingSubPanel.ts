import { t } from '@lingui/macro'
import {
  useProjectContext,
  useProjectMetadata,
} from 'components/ProjectDashboard/hooks'
import { useFundingCycleCountdown } from 'components/ProjectDashboard/hooks/useFundingCycleCountdown'
import { timeSecondsToDateString } from 'components/ProjectDashboard/utils/timeSecondsToDateString'
import { useProjectUpcomingFundingCycle } from 'hooks/v2v3/contractReader/useProjectUpcomingFundingCycle'
import { useMemo } from 'react'

export const useCurrentUpcomingSubPanel = (type: 'current' | 'upcoming') => {
  const { projectId } = useProjectMetadata()
  const {
    fundingCycle,
    loading: { fundingCycleLoading },
  } = useProjectContext()
  const {
    data: upcomingFundingCycleData,
    loading: upcomingFundingCycleLoading,
  } = useProjectUpcomingFundingCycle({ projectId })
  const [upcomingFundingCycle] = upcomingFundingCycleData ?? []
  const { timeRemainingText } = useFundingCycleCountdown()

  const cycleNumber = useMemo(() => {
    if (type === 'current') {
      return fundingCycle?.number.toNumber()
    }
    return fundingCycle?.number ? fundingCycle.number.toNumber() + 1 : undefined
  }, [fundingCycle?.number, type])

  const cycleUnlocked = useMemo(() => {
    if (type === 'current') {
      return fundingCycle?.duration?.isZero() ?? true
    }
    return upcomingFundingCycle?.duration?.isZero() ?? true
  }, [fundingCycle?.duration, type, upcomingFundingCycle?.duration])

  const upcomingCycleLength = useMemo(() => {
    if (!upcomingFundingCycle) return
    if (cycleUnlocked) return '-'
    return timeSecondsToDateString(
      upcomingFundingCycle.duration.toNumber(),
      'short',
    )
  }, [cycleUnlocked, upcomingFundingCycle])

  const status = useMemo(() => {
    if (cycleUnlocked) return t`Unlocked`
    return t`Locked`
  }, [cycleUnlocked])

  const remainingTime = useMemo(() => {
    if (cycleUnlocked) return '-'
    return timeRemainingText
  }, [cycleUnlocked, timeRemainingText])

  // Short circuit current for faster loading
  if (type === 'current') {
    if (fundingCycleLoading) return { loading: true, type }
    return {
      loading: false,
      type,
      cycleNumber,
      status,
      remainingTime,
    }
  }

  if (fundingCycleLoading || upcomingFundingCycleLoading)
    return {
      loading: true,
      type,
    }

  return {
    loading: false,
    type,
    cycleNumber,
    status,
    cycleLength: upcomingCycleLength,
  }
}
