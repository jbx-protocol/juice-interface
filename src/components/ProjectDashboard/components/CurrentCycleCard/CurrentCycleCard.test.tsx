/**
 * @jest-environment jsdom
 */

import { render } from '@testing-library/react'
import { useCurrentCycleCard } from 'components/ProjectDashboard/hooks'
import { useProjectPageQueries } from 'components/ProjectDashboard/hooks/useProjectPageQueries'
import { CurrentCycleCard } from './CurrentCycleCard'

jest.mock('components/ProjectDashboard/hooks')
jest.mock('components/ProjectDashboard/hooks/useProjectPageQueries')

describe('CurrentCycleCard', () => {
  const DefaultUseCurrentCycleCardMock = {
    currentCycleNumber: 1,
    timeRemainingText: '1 day',
  }
  beforeEach(() => {
    ;(useCurrentCycleCard as jest.Mock).mockReturnValue(
      DefaultUseCurrentCycleCardMock,
    )
    ;(useProjectPageQueries as jest.Mock).mockReturnValue({
      setProjectPageTab: jest.fn(),
    })
  })

  it('renders', () => {
    const { container } = render(<CurrentCycleCard />)
    expect(container).toMatchSnapshot()
  })
})
