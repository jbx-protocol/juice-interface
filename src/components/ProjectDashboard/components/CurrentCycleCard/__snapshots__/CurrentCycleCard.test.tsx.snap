// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CurrentCycleCard renders 1`] = `
<div>
  <div
    class="rounded-lg bg-smoke-50 py-5 px-6 dark:bg-slate-700 min-w-0 cursor-pointer pr-9"
  >
    <span
      class="inline-flex items-center gap-1"
    >
      <span
        class="truncate whitespace-nowrap text-base font-medium"
      >
        Current Cycle: #1
      </span>
      <svg
        aria-hidden="true"
        class="h-4 w-4 text-grey-500 dark:text-slate-200"
        fill="none"
        stroke="currentColor"
        stroke-width="1.5"
        viewBox="0 0 24 24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M9.879 7.519c1.171-1.025 3.071-1.025 4.242 0 1.172 1.025 1.172 2.687 0 3.712-.203.179-.43.326-.67.442-.745.361-1.45.999-1.45 1.827v.75M21 12a9 9 0 11-18 0 9 9 0 0118 0zm-9 5.25h.008v.008H12v-.008z"
          stroke-linecap="round"
          stroke-linejoin="round"
        />
      </svg>
    </span>
    <div
      class="mt-4 flex min-w-0 items-center gap-2"
    >
      <div
        aria-label=""
        class="truncate text-2xl font-medium"
      >
        1 day
      </div>
      <svg
        aria-hidden="true"
        class="h-4 w-4 flex-shrink-0 text-smoke-400 dark:text-slate-300"
        fill="currentColor"
        viewBox="0 0 24 24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          clip-rule="evenodd"
          d="M12 1.5a5.25 5.25 0 00-5.25 5.25v3a3 3 0 00-3 3v6.75a3 3 0 003 3h10.5a3 3 0 003-3v-6.75a3 3 0 00-3-3v-3c0-2.9-2.35-5.25-5.25-5.25zm3.75 8.25v-3a3.75 3.75 0 10-7.5 0v3h7.5z"
          fill-rule="evenodd"
        />
      </svg>
    </div>
    <div
      aria-label=""
      class="truncate mt-2 text-smoke-500 dark:text-slate-200"
    >
      until Cycle #2
    </div>
  </div>
</div>
`;
